runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -376,6 +376,63 @@ function M.get(bufnr, client_id)
   return diagnostic_cache[bufnr][client_id] or {}
 end
 
+--- Get the diagnostics by position
+---
+---@param bufnr number The buffer number
+---@param position table The (0,0)-indexed position
+---@param opts table|nil Configuration keys
+---         - severity: (DiagnosticSeverity, default nil)
+---             - Only return diagnostics with this severity. Overrides severity_limit
+---         - severity_limit: (DiagnosticSeverity, default nil)
+---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
+---@param client_id number the client id
+---@return table Table with map of line number to list of diagnostics.
+--               Structured: { [1] = {...}, [5] = {.... } }
+function M.get_position_diagnostics(bufnr, position, opts, client_id)
```

Comment:
I wonder if we should instead extend `get` with a predicate function.

With `get`, `get_all`, `get_line_diagnostics` and now `get_position_diagnostics` it looks as if there's a trend to create a new function for each different set of diagnostics somebody wants to retrieve.

Reply:
+1, I initially suggested changes to get_position_diagnostics but I think it's way better to consolidate things and pass a predicate function depending on what slice of diagnostics you want

Reply:
> we should instead extend get with a predicate function
i'm not sure what is being suggested here, can you give an example of how the consolidated function will look like?

Reply:
For example, `M.get_all` could take an optional `predicate` as argument, that if present can be used to filter the arguments included.

You'd then call it within `show_position_diagnostics` with a predicate that filters the diagnostics by position:

```lua
local position_diagnostics = M.get_all(nil, function(diag)
  return (position[2] >= diag.range['start'].character)
    and (position[2] <= diag.range['end'].character or position[1] < diag.range['end'].line)
end)
```

Reply:
:+1: i tried to move the diagnostic retrieval logic for position diagnostics into `get`, what do you think?